### ISSUE TITLE
am/gpu: need buffer pool when allocating staging buffer by using MPL_gpu_malloc_host

### ISSUE THREAD

**Issue by minsii** [pmodels/mpich#5171]
**Problem description**: In current CH4 AM code, the host staging buffer is allocated by using `MPL_gpu_malloc_host` at each send/recv call (e.g., in UCX netmod `MPIDI_NM_am_isend` and `MPIDI_NM_progress` for send/recv side copy, respectively). When CUDA is enabled, it internally calls `cudaMallocHost`, at each send/recv.

The benefit to use `cudaMallocHost` is to leverage fast DMA based data transfer, however, the overhead of `cudaMallocHost` is so expensive that the cost of (`cudaMallocHost + dma`) can be 1000x more expensive than that of (`malloc+slow-copy`).  See profiling results below (similar results observed on Summit, with cuda-11.2).

**Test program**: OSU mpi/pt2pt/osu_latency -d cuda D D (pingpong test, thus `latency = send_cost + recv_cost ~= (cudaMallocHost + dma) * 2 + network_latency`
**Configuration**: MPICH/main, `--with-device=ch4:ucx --enable-ch4-am-only` (`DIRECT_NETMOD` is enabled by default)
**Platform**: jlse/GPU_V100_SMX2, cuda-10.0.130, single node with `-np 2 -ppn 2`

**Profiling results with default code** (`am_send_alloc_pack_time = cudaMallocHost + MPIR_Typerep_pack`, `am_progress_recv_time = cudaMallocHost + ucp_tag_msg_recv_nb` )
```
# Size          Latency (us)
Rank 1 cudaSetDevice 1/4
0                    2331.18
0       , am_send_alloc_pack_time=1038.27,  am_progress_recv_time=1070.60
1                    2389.79
1       , am_send_alloc_pack_time=1069.00,  am_progress_recv_time=1097.09
2                    2387.53
2       , am_send_alloc_pack_time=1067.63,  am_progress_recv_time=1094.81
4                    2385.85
4       , am_send_alloc_pack_time=1065.77,  am_progress_recv_time=1094.78
8                    2389.16
8       , am_send_alloc_pack_time=1067.17,  am_progress_recv_time=1094.42
16                   2388.88
16      , am_send_alloc_pack_time=1065.36,  am_progress_recv_time=1094.96
32                   2390.41
32      , am_send_alloc_pack_time=1067.46,  am_progress_recv_time=1094.99
64                   2393.11
64      , am_send_alloc_pack_time=1066.40,  am_progress_recv_time=1095.86
128                  2392.40
128     , am_send_alloc_pack_time=1065.91,  am_progress_recv_time=1093.99
256                  2397.81
256     , am_send_alloc_pack_time=1066.93,  am_progress_recv_time=1096.57
512                  2398.46
512     , am_send_alloc_pack_time=1066.94,  am_progress_recv_time=1094.79
1024                 2401.83
1024    , am_send_alloc_pack_time=1070.00,  am_progress_recv_time=1100.39
2048                 2400.38
2048    , am_send_alloc_pack_time=1068.72,  am_progress_recv_time=1097.61
4096                 2401.03
4096    , am_send_alloc_pack_time=1067.78,  am_progress_recv_time=1099.30
8192                 2405.07
8192    , am_send_alloc_pack_time=1069.32,  am_progress_recv_time=1098.93
```

**Profiling results by replacing cudaMallocHost with malloc**
```
# Size          Latency (us)
0                       1.50
0   , am_send_alloc_pack_time=0.04,  am_progress_recv_time=0.35
1                      40.85
1   , am_send_alloc_pack_time=25.92,  am_progress_recv_time=20.14
2                      36.86
2   , am_send_alloc_pack_time=16.84,  am_progress_recv_time=17.47
4                      35.29
4   , am_send_alloc_pack_time=16.81,  am_progress_recv_time=17.53
8                      35.35
8   , am_send_alloc_pack_time=16.71,  am_progress_recv_time=16.68
16                     35.81
16  , am_send_alloc_pack_time=17.03,  am_progress_recv_time=16.99
32                     34.67
32  , am_send_alloc_pack_time=16.78,  am_progress_recv_time=16.51
64                     35.67
64  , am_send_alloc_pack_time=16.85,  am_progress_recv_time=16.71
128                    35.03
128 , am_send_alloc_pack_time=16.76,  am_progress_recv_time=16.69
256                    35.11
256 , am_send_alloc_pack_time=16.82,  am_progress_recv_time=16.77
512                    35.25
512 , am_send_alloc_pack_time=16.92,  am_progress_recv_time=16.83
1024                   36.02
1024    , am_send_alloc_pack_time=16.97,  am_progress_recv_time=17.49
2048                   36.43
2048    , am_send_alloc_pack_time=17.18,  am_progress_recv_time=17.52
4096                   37.93
4096    , am_send_alloc_pack_time=17.19,  am_progress_recv_time=18.90
8192                   40.58
8192    , am_send_alloc_pack_time=17.50,  am_progress_recv_time=20.21
```

**Note**: the above results are 2 processes on single node, with device buffer on both sender/receiver. Results with 2 nodes, device-host, host-device, host-host are similar.

**Solutions**
- [x] Temporarily disable `cudaMallocHost` in `MPL_gpu_malloc_host` (not sure performance with ZE)
- [ ] Implement memory pool for `cudaMallocHost` staging buffers

**Comment by hzhou** [comment]
The memory pool may be difficult to implement -- there isn't a fixed size for these messages, right? It kinda makes sense that we should not do these optimizations for one-time buffers. Anything we can do for one-time buffers, CUDA internally can always do better. I think we were just committing a case of pre-mature optimization.

**Comment by minsii** [comment]
@yfguo suggested that the overhead might go away once we apply genq to ucx am.

To @hzhou 's question, the overhead of creation the one-time buffer is about 1000ms. This is no pre-mature optimization in CUDA. It is like `mmap` -- always goes into kernel and maps the pages, no user space pool. Based on google results, CUDA recommends user maintains preallocation of these buffers to mitigate the creation overhead

**Comment by yfguo** [comment]
Just putting my notes here. What we did for OFI is getting pack buffer from a preallocated pool for non-contig DT or GPU buffers. The implementation uses GenQ private pool and the pipeline. In UCX netmod, we still use the gpu_malloc_host to allocate the pack buffer at send time which is painfully slow. One solution to this is to employ the same GenQ+pipeline design.

**Comment by minsii** [comment]
@yfguo @hzhou Can we have an estimation when GenQ+pipeline can be applied for UCX? A workable PR is OK for me to continue evaluation, as long as isend/irecv and RMA work fine.

**Comment by hzhou** [comment]
> Temporarily disable cudaMallocHost in MPL_gpu_malloc_host (not sure performance with ZE)

@minsii Wouldn't this allow you to continue evaluation?

**Comment by minsii** [comment]
> > Temporarily disable cudaMallocHost in MPL_gpu_malloc_host (not sure performance with ZE)
> 
> @minsii Wouldn't this allow you to continue evaluation?

Yes, I can make a quick workaround branch following this way. It at least can give me reasonable performance. But I would prefer a more optimal version (i.e., still getting benefit from `cudaMallocHost` for large data transfer via DMA while eliminating malloc overhead by using the preallocated pool). I think that the latency of large msg size can be further reduced by the latter approach.

Given said that, if you see that it may take more time (e.g., a few weeks or more) to make the  GenQ+pipeline for UCX PR, I will use the former workaround for oshmpi Q1Q2 milestone.

**Comment by yfguo** [comment]
The UCX AM pipeline PR is a temporary solution for this. We probably eventually want to use the GPU support in UCX AM directly (https://github.com/openucx/ucx/pull/6041), which why the merge for UCX AM pipeline is on halt.

**Comment by hzhou** [comment]
Closing this issue since we merged #6074 and we replaced all `MPIR_gpu_malloc_host` with `MPL_malloc`. It is still worth pursuing to use genq pool and pipelining. Moving this issue into internal todo.
